design/src/rfcs/rfc0031_providing_fallback_credentials_on_timeout.md

```diff
@@ -0,0 +1,251 @@
+        future::ProvideCredentials::new(async move {
+            let credentials = future
+                .await
+                .map_err(|_err| CredentialsError::provider_timed_out(timeout))?;
+            credentials
+        })
+    }
+```
+`provide_credentials_with_timeout` encapsulated the timeout race and allowed users to specify how long the external timeout for `provide_credentials` would be. The code snippet from `LazyCredentialsCache` then looked like
+```rust
+let sleeper = Arc::clone(&self.sleeper);
+let load_timeout = self.load_timeout; // by default self.load_timeout is 5 seconds.
+// --snip--
+let result = cache
+    .get_or_load(|| {
+        async move {
+            let credentials = provider
+                .provide_credentials_with_timeout(sleeper, load_timeout)
+                .await?;
+            // --snip--
+        }
+    }).await;
+// --snip--
+```
+However, implementing `CredentialsProviderChain::provide_credentials_with_timeout` quickly ran into the following problem:
+```rust
+impl ProvideCredentials for CredentialsProviderChain {
+    // --snip--
+
+    fn provide_credentials_with_timeout<'a>(
+        &'a self,
+        sleeper: Arc<dyn AsyncSleep>,
+        timeout: Duration,
+    ) -> future::ProvideCredentials<'a>
+    where
+        Self: 'a,
+    {
+        future::ProvideCredentials::new(self.credentials_with_timeout(sleeper, timeout))
+    }
+}
+
+impl CredentialsProviderChain {
+    // --snip--
+
+    async fn credentials_with_timeout(
+        &self,
+        sleeper: Arc<dyn AsyncSleep>,
+        timeout: Duration,
+    ) -> provider::Result {
+        for (_, provider) in &self.providers {
+            match provider
+                .provide_credentials_with_timeout(Arc::clone(&sleeper), /* how do we calculate timeout for each provider ? */)
+                .await
+            {
+                Ok(credentials) => {
+                    return Ok(credentials);
+                }
+                Err(CredentialsError::ProviderTimedOut(_)) => {
+                    // --snip--
+                }
+                Err(err) => {
+                   // --snip--
+                }
+           }
+        }
+        Err(CredentialsError::provider_timed_out(timeout))
+    }
+```
+There are mainly two problems with this approach. The first problem is that as shown above, there is no sensible way to calculate a timeout for each provider in the chain. The second problem is that exposing a parameter like `timeout` at a public trait's level is giving too much control to users; delegating overall timeout to the individual provider means each provider has to get it right.
+
+Changes checklist
+-----------------
+
+- [ ] Add `on_timeout` method to the `ProvideCredentials` trait with the default implementation
+- [ ] Implement the `OnTimeout` future newtype
+- [ ] Implement `CredentialsProviderChain::on_timeout`
+- [ ] Add unit tests for `Case 1` and `Case 2`
```

Comment:
When new credentials come through, we should test that all layer's cached credentials are refreshed.

Reply:
Thanks for coming up with the use case! As discussed, we can employ a different approach. See the [Appendix](https://github.com/awslabs/smithy-rs/blob/d413e58e57dc0fd7bcd68583414bb7aa250c684c/design/src/rfcs/rfc0031_providing_fallback_credentials_on_timeout.md#appendix) section.